lkmpg.tex

```diff
@@ -1554,6 +1554,97 @@ \section{System Calls}
+It is usually used as an event tracer by attaching callbacks to the beginning of functions in order to record and trace the flow of the kernel.
+
+The basic prototype of the callback function is
+\begin{verbatim}
+struct ftrace_ops {
+  ftrace_func_t func; // callback function
+  unsigned long flags; // ftrace flags
+  void* private; // any private data
+};
+void callback_func(unsigned long ip, unsigned long parent_ip,
+                   struct ftrace_ops *ops, struct pt_regs *regs);
+\end{verbatim}
+
+where
+
+\begin{itemize}
+  \item \cpp|ip|: The instruction pointer of the function being traced.
+  \item \cpp|parent_ip|: The instruction pointer of the caller of the traced function.
+  \item \cpp|ops|: A pointer to \cpp|ftrace_ops| that was used to register the callback.
+  \item \cpp|regs|: If \cpp|FTRACE_OPS_FL_SAVE_REGS| or \cpp|FTRACE_OPS_FL_SAVE_REGS_IF_SUPPORTED| are set in the \cpp|ftrace_ops| structure, then this will be pointing to the \cpp|pt_regs| structure like it would be if an breakpoint was placed at the start of the function where \verb|ftrace| was tracing for CPU register access. Otherwise it either contains garbage, or \cpp|NULL|. Do notice that in kernel version later than v5.11, this is replaced with \cpp|struct ftrace_regs *fregs|, with the original \cpp|pt_regs| accessible by \cpp|fregs->regs|.
+\end{itemize}
+
+Internally, there's a 5-byte \cpp|call| to \cpp|__fentry__| at the beginning (BEFORE function prologue) of a traceable kernel function, which is converted to \cpp|nop| during boot to prevent overhead. When a trace is registered, it is changed back to \cpp|__fentry__| and the registered callback will be executed accordingly.
+
+But callbacks can do more.
+Since it's called at the start of a function,
+and we have access to CPU registers,
+maybe we can ``hijack'' the traced function by modifying the instruction pointer?
+Yes, this is possible by enabling \cpp|FTRACE_OPS_FL_IPMODIFY| flag when registering a trace.
+It will allow us to modify the instruction pointer register, which will become an unconditional jump after the \verb|ftrace| function.
+Note that while there can be multiple tracers on one function, only one tracer that changes \cpp|ip| can be registered at the same time.
+
+Figure~\ref{img:ftrace-hooking-example} gives an example of auditing \cpp|sys_execve| by hooking it using \verb|ftrace|.
+The callback function (\cpp|fh_ftrace_thunk|) checks whether the call is from the kernel or the module,
+and passes control accordingly.
+If the call is from the kernel, our auditing function is called.
+Otherwise, nothing happens.
+The check is important because we're only ``decorating'' the original syscall.
+Our auditing function contains call to the original \cpp|sys_execve|,
+which will trigger the callback function again.
+It'll be an infinite loop if there's no check performed.
+
+\begin{figure}[h]
+  \centering
+  \includegraphics[width=\textwidth]{assets/syscall/flow.jpg}
```

Comment:
Utilize TikZ for drawing. See https://texample.net/tikz/examples/pgf-umlsd/
Avoid putting bitmap files.

Reply:
Also, I have to rotate the figure by 90 degrees to minimize the overflow ~~although it overflows anyway~~. Will it affect the output of the website or should I rotate it back and let it overflow?

Reply:
Alternatively, you can just render the partial sequences.

Reply:
I don't think I can rework this diagram in latex because it lacks features I need, like returning to functions other than caller (this is the most important one), annotation. (I'm not the creator of this sequence diagram so obtaining the "original" file is not possible either.)

I think the resolution of that jpeg is good enough for even printing, so I think I may left it as is.

<details><summary>The latex code I've written and the result</summary>

```tex
\begin{sequencediagram}
  \newthread{do_syscall_64}{do\_syscall\_64}
  \newinst[1.5]{sys_execve}{sys\_execve}
  \newinst[1.5]{ftrace}{[ftrace]}
  \newinst[1]{fh_ftrace_thunk}{fh\_ftrace\_thunk}
  \newinst[1]{fh_sys_execve}{fh\_sys\_execve}

      \postlevel \postlevel \postlevel
      \begin{call}{do_syscall_64}{\shortstack{
        \cpp|regs-ax=|\\
        \cpp|sys_call_table[nr]|\\
        \cpp|(regs->di,regs->si|\\
        \cpp|regs->dx,regs->r10|\\
        \cpp|regs->r8,regs->r9)|
      }}{sys_execve}{}
        \begin{call}{sys_execve}{call \cpp|__fentry__|}{ftrace}{}
          \begin{call}{ftrace}{}{fh_ftrace_thunk}{}
            \postlevel
          \end{call}
        \end{call}
        \begin{call}{sys_execve}{hooking}{fh_sys_execve}{\cpp|real_sys_execve()|}
          \postlevel
        \end{call}
        \postlevel
        \begin{call}{sys_execve}{call \cpp|__fentry__|}{ftrace}{}
          \begin{call}{ftrace}{}{fh_ftrace_thunk}{}
            \postlevel
          \end{call}
        \end{call}
        \begin{call}{sys_execve}{}{fh_sys_execve}{}
        \end{call}
      \end{call}
\end{sequencediagram}
```

![image](https://user-images.githubusercontent.com/45678482/205811001-4869abfb-baf3-4d79-804d-911bf444f39f.png)

</details>